scrapy/core/http2/protocol.py

```diff
@@ -0,0 +1,204 @@
+        return stream
+
+    def _send_pending_requests(self):
+        # TODO: handle MAX_CONCURRENT_STREAMS
+        # Initiate all pending requests
+        while self._pending_request_stream_pool:
+            stream = self._pending_request_stream_pool.popleft()
+            stream.initiate_request()
+
+    def _write_to_transport(self):
+        """ Write data to the underlying transport connection
+        from the HTTP2 connection instance if any
+        """
+        data = self.conn.data_to_send()
+        self.transport.write(data)
+
+        logger.debug("Sent {} bytes to {} via transport".format(len(data), self._metadata['ip_address']))
+
+    def request(self, request: Request):
+        stream = self._new_stream(request)
+        d = stream.get_response()
+
+        # If connection is not yet established then add the
+        # stream to pool or initiate request
+        if self.is_connection_made:
+            stream.initiate_request()
+        else:
+            self._pending_request_stream_pool.append(stream)
+
+        return d
+
+    def connectionMade(self):
+        """Called by Twisted when the connection is established. We can start
+        sending some data now: we should open with the connection preamble.
+        """
+        self.destination = self.transport.getPeer()
+        logger.info('Connection made to {}'.format(self.destination))
+
+        self._metadata['certificate'] = Certificate(self.transport.getPeerCertificate())
+        self._metadata['ip_address'] = ipaddress.ip_address(self.destination.host)
+
+        self.conn.initiate_connection()
+        self._write_to_transport()
+        self.is_connection_made = True
+
+    def dataReceived(self, data):
+        try:
+            events = self.conn.receive_data(data)
+            self._handle_events(events)
+        except ProtocolError as e:
+            # TODO: In case of InvalidBodyLengthError -- terminate only one stream
+
+            # Save this error as ultimately the connection will be dropped
+            # internally by hyper-h2. Saved error will be passed to all the streams
+            # closed with the connection.
+            self._protocol_error = e
+
+            # We lose the transport connection here
+            self.transport.loseConnection()
+        finally:
+            self._write_to_transport()
+
+    def connectionLost(self, reason=connectionDone):
+        """Called by Twisted when the transport connection is lost.
+        No need to write anything to transport here.
+        """
+        # Pop all streams which were pending and were not yet started
+        # NOTE: Stream.close() pops the element from the streams dictionary
+        # which raises `RuntimeError: dictionary changed size during iteration`
+        # Hence, we copy the streams into a list.
+        for stream in list(self.streams.values()):
+            stream.close(StreamCloseReason.CONNECTION_LOST, self._protocol_error)
```

Comment:
@adityaa30 @Gallaecio I still don't like the fact that a Stream ends up being responsible from removing itself from streams dict, and that this implementation details demands the need to create a new structure here. What about this?
```suggestion
        for stream in self.streams.values():
            stream.close(StreamCloseReason.CONNECTION_LOST, self._protocol_error)
        self.streams.clear()
```
Since all streams will be popped anyway.

```

Reply:
This looks good. I have one doubt -- the `Stream` class under differnet conditions closes itself. Eg. the client cancels the request by calling `response_defer.cancel()`. For such cases what should we do? :thinking:

Reply:
Interesting. In that case, perhaps the stream could hold a reference to its parent protocol. It already stores a `H2ConnectionMetadataDict` with metadata, it could store the protocol itself instead of it.